File: packages/sdk-go/conformance/errors_test.go
package conformance

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/gemini/developer-platform/packages/sdk-go/transport"
)

func TestErrorMapping(t *testing.T) {
	manifest := LoadManifest(t)
	var suite FixtureSuite
	for _, candidate := range manifest.Suites {
		if candidate.ID == "http/errors" {
			suite = candidate
			break
		}
	}
	if suite.ID == "" {
		t.Fatal("manifest does not contain http/errors suite")
	}
	for _, caseID := range suite.Cases {
		t.Run(caseID, func(t *testing.T) {
			var fixture FixtureError
			if err := json.Unmarshal(LoadCase(t, suite.ID, caseID), &fixture); err != nil {
				t.Fatalf("decoding fixture: %v", err)
			}
			if fixture.ID != suite.ID+"/"+caseID || fixture.Kind != suite.Kind {
				t.Fatalf("fixture identity = (%q, %q), want (%q, %q)", fixture.ID, fixture.Kind, suite.ID+"/"+caseID, suite.Kind)
			}
			testErrorCase(t, fixture)
		})
	}
}

func testErrorCase(t *testing.T, fixture FixtureError) {
	t.Helper()
	server := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		if req.Method != http.MethodGet {
			t.Errorf("handler method = %s, want GET", req.Method)
		}
		for name, value := range fixture.Response.Headers {
			w.Header().Set(name, value)
		}
		w.WriteHeader(fixture.Response.Status)
		if fixture.Response.Body != "" {
			_, _ = io.WriteString(w, fixture.Response.Body)
		}
	}))
	defer server.Close()

	client := transport.NewClient(
		transport.WithHTTPClient(server.Client()),
		transport.WithRetryPolicy(transport.RetryPolicy{MaxRetries: 0}),
	)
	_, _, err := client.RequestRaw(context.Background(), http.MethodGet, server.URL+"/v1/conformance", nil)
	if err == nil {
		t.Fatal("expected fixture response to return an error")
	}
	wantSentinel := errorForCanonicalKind(fixture.Expect.Kind)
	if wantSentinel == nil {
		t.Fatalf("unsupported canonical error kind %q", fixture.Expect.Kind)
	}
	if !errors.Is(err, wantSentinel) {
		t.Errorf("error %v does not match %v for canonical kind %q", err, wantSentinel, fixture.Expect.Kind)
	}

	var apiErr *transport.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("error %T does not expose *transport.APIError", err)
	}
	if apiErr.StatusCode != fixture.Response.Status {
		t.Errorf("error status = %d, want %d", apiErr.StatusCode, fixture.Response.Status)
	}
	if string(apiErr.RawBody) != fixture.Response.Body {
		t.Errorf("error raw body = %q, want exact %q", apiErr.RawBody, fixture.Response.Body)
	}
	if fixture.Expect.Reason != "" {
		reason := apiErr.Reason
		if reason == "" {
			reason = apiErr.ErrorMessage
		}
		if reason != fixture.Expect.Reason {
			t.Errorf("error reason = %q, want %q", reason, fixture.Expect.Reason)
		}
	}
	if fixture.Expect.RetryAfterSeconds != nil {
		var rateErr *transport.RateLimitError
		if !errors.As(err, &rateErr) {
			t.Fatalf("error %T does not expose *transport.RateLimitError", err)
		}
		want := time.Duration(*fixture.Expect.RetryAfterSeconds) * time.Second
		if rateErr.RetryAfter != want {
			t.Errorf("retry-after = %v, want %v", rateErr.RetryAfter, want)
		}
	}
	if fixture.Response.Body == "" && len(apiErr.RawBody) != 0 {
		t.Errorf("empty response body produced raw bytes %q", apiErr.RawBody)
	}
}

func errorForCanonicalKind(kind string) error {
	switch kind {
	case "invalid_nonce":
		return transport.ErrInvalidNonce
	case "missing_nonce":
		return transport.ErrMissingNonce
	case "invalid_signature":
		return transport.ErrInvalidSignature
	case "missing_role":
		return transport.ErrMissingRole
	case "terms_required":
		return transport.ErrAcceptTermsRequired
	case "insufficient_funds":
		return transport.ErrInsufficientFunds
	case "rate_limited":
		return transport.ErrRateLimited
	case "order_not_found":
		return transport.ErrOrderNotFound
	case "market_closed":
		return transport.ErrMarketClosed
	case "not_found":
		return transport.ErrNotFound
	case "service_error":
		return transport.ErrInternalServer
	case "invalid_request":
		return transport.ErrBadRequest
	default:
		return nil
	}
}

File: packages/sdk-typescript/src/tests/conformance/errors.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import {
  AcceptTermsRequired,
  ApiError,
  InsufficientFunds,
  InvalidNonce,
  InvalidRequest,
  InvalidSignature,
  MissingNonce,
  MissingRole,
  NotFoundError,
  RateLimitError,
  ServiceUnavailable,
  serializeError,
} from "../../errors.js";
import { HttpTransport, parseLosslessJson } from "../../transport/http.js";
import { streamingTextResponse } from "../support/http-fixtures.js";
import {
  loadCase,
  loadManifest,
  type FixtureCase,
} from "./support/fixtures.js";

type ErrorKind =
  | "invalid_nonce"
  | "missing_nonce"
  | "invalid_signature"
  | "missing_role"
  | "terms_required"
  | "insufficient_funds"
  | "rate_limited"
  | "order_not_found"
  | "market_closed"
  | "not_found"
  | "service_error"
  | "invalid_request";
type ErrorFixture = FixtureCase & {
  readonly response: {
    readonly status: number;
    readonly headers?: Record<string, string>;
    readonly body: string;
  };
  readonly expect: {
    readonly kind: ErrorKind;
    readonly reason?: string;
    readonly retryAfterSeconds?: number;
  };
};

const CLASS_BY_KIND: Record<ErrorKind, typeof ApiError> = {
  invalid_nonce: InvalidNonce,
  missing_nonce: MissingNonce,
  invalid_signature: InvalidSignature,
  missing_role: MissingRole,
  terms_required: AcceptTermsRequired,
  insufficient_funds: InsufficientFunds,
  rate_limited: RateLimitError,
  order_not_found: NotFoundError,
  market_closed: ApiError,
  not_found: NotFoundError,
  service_error: ServiceUnavailable,
  invalid_request: InvalidRequest,
};

function fixture(value: FixtureCase): ErrorFixture {
  return value as unknown as ErrorFixture;
}

function responseHeaders(values: Record<string, string> | undefined): { get(name: string): string | null } {
  const entries = Object.entries(values ?? {});
  return {
    get(name) {
      const entry = entries.find(([key]) => key.toLowerCase() === name.toLowerCase());
      return entry?.[1] ?? null;
    },
  };
}

async function runCase(value: ErrorFixture): Promise<void> {
  const transport = new HttpTransport({
    env: "sandbox",
    maxRetries: 0,
    fetchImpl: async () => streamingTextResponse(
      value.response.body,
      value.response.status,
      responseHeaders(value.response.headers),
    ),
  });
  let observed: unknown;
  try {
    await transport.requestPublic({ method: "GET", path: "/v1/conformance" });
  } catch (error) {
    observed = error;
  }
  assert.ok(observed instanceof ApiError, "HTTP error must be an ApiError");
  assert.equal(observed.status, value.response.status, "HTTP status must be preserved");
  const expectedClass = CLASS_BY_KIND[value.expect.kind];
  assert.ok(observed instanceof expectedClass, `expected ${value.expect.kind} to map to ${expectedClass.name}`);
  if (value.expect.reason !== undefined) assert.equal(observed.reason, value.expect.reason);
  if (value.expect.retryAfterSeconds !== undefined) {
    assert.equal(
      observed.metadata?.rateLimit?.retryAfter,
      String(value.expect.retryAfterSeconds),
    );
  }

  const serialized = serializeError(observed, { includeRawBody: true });
  if (value.response.body.length === 0) {
    assert.equal(serialized.body, undefined);
  } else {
    let expectedBody: unknown = value.response.body;
    try {
      expectedBody = parseLosslessJson(value.response.body);
    } catch {
      // Keep unstructured response text as the retained raw body.
    }
    assert.deepEqual(serialized.body, expectedBody);
  }
}

const manifest = await loadManifest();
const suite = manifest.suites.find((entry) => entry.id === "http/errors");
if (!suite) throw new Error("conformance manifest is missing http/errors");
for (const caseId of suite.cases) {
  test(`conformance error mapping: ${caseId}`, async () => {
    await runCase(fixture(await loadCase(suite.id, caseId)));
  });
}
